### PR TITLE
[InstSimplify] Do not simplify freeze in `simplifyWithOpReplaced`

### LINKED ISSUE #91178: [InstCombine] Miscompilation which simplifies a use of a multi-use freeze
Reduced test case: https://alive2.llvm.org/ce/z/_JyUGw
```
define i8 @func78(i1 %0, i8 %1) {
  %3 = zext i1 %0 to i8
  %4 = shl i8 %3, %1
  %5 = freeze i8 %4
  %6 = select i1 %0, i8 0, i8 %5
  %7 = call i8 @llvm.fshr.i8(i8 0, i8 %5, i8 %6)
  ret i8 %7
}
```

```
----------------------------------------
define i8 @func78(i1 %#0, i8 %#1) {
#2:
  %#3 = zext i1 %#0 to i8
  %#4 = shl i8 %#3, %#1
  %#5 = freeze i8 %#4
  %#6 = select i1 %#0, i8 0, i8 %#5
  %#7 = fshr i8 0, i8 %#5, i8 %#6
  ret i8 %#7
}
=>
define i8 @func78(i1 %#0, i8 %#1) nofree noundef willreturn memory(none) {
#2:
  %#3 = zext i1 %#0 to i8
  %#4 = shl nuw i8 %#3, %#1
  %#5 = freeze i8 %#4
  ret i8 %#5
}
Transformation doesn't verify!

ERROR: Value mismatch

Example:
i1 %#0 = #x0 (0)
i8 %#1 = #x08 (8)

Source:
i8 %#3 = #x00 (0)
i8 %#4 = poison
i8 %#5 = #x00 (0)
i8 %#6 = #x00 (0)
i8 %#7 = #x00 (0)

Target:
i8 %#3 = #x00 (0)
i8 %#4 = poison
i8 %#5 = #x9d (157, -99)
Source value: #x00 (0)
Target value: #x9d (157, -99)

Summary:
  0 correct transformations
  1 incorrect transformations
  0 failed-to-prove transformations
  0 Alive2 errors
```

### PR DESCRIPTION
**Pull request by dtcxzyw**
See the LangRef:
> All uses of a value returned by the same ‘freeze’ instruction are guaranteed to always observe the same value, while different ‘freeze’ instructions may yield different values.

It is incorrect to replace freezes with the simplified value.

Proof:
https://alive2.llvm.org/ce/z/3Dn9Cd
https://alive2.llvm.org/ce/z/Qyh5h6

Fixes https://github.com/llvm/llvm-project/issues/91178

### CODE REVIEW

File: llvm/lib/Analysis/InstructionSimplify.cpp [hunk 1]
```diff
@@ -4312,6 +4312,10 @@ static Value *simplifyWithOpReplaced(Value *V, Value *Op, Value *RepOp,
   if (match(I, m_Intrinsic<Intrinsic::is_constant>()))
     return nullptr;
 
+  // Don't simplify a use of a multi-use freeze.
+  if (isa<FreezeInst>(I) && !I->hasOneUse())
```
Comment (nikic) [inline llvm/lib/Analysis/InstructionSimplify.cpp:4316]: We cannot check uses in InstSimplify. There could be multi-use down-stream and not all of them replaced, or similar. We should just skip freeze entirely.

**Comment by dtcxzyw** [comment]
Another case with one-use freeze: https://alive2.llvm.org/ce/z/Qyh5h6
```
define i1 @func638(i16 %0, i16 %1) {
  %3 = icmp uge i16 %0, %1
  %4 = sext i1 %3 to i16
  %5 = freeze i16 %4
  %6 = icmp uge i16 %5, %1
  ret i1 %6
}
```
```

----------------------------------------
define i1 @func638(i16 %#0, i16 %#1) {
#2:
  %#3 = icmp uge i16 %#0, %#1
  %#4 = sext i1 %#3 to i16
  %#5 = freeze i16 %#4
  %#6 = icmp uge i16 %#5, %#1
  ret i1 %#6
}
=>
define i1 @func638(i16 %#0, i16 %#1) nofree noundef willreturn memory(none) {
#2:
  %#3 = icmp uge i16 %#0, %#1
  %.fr = freeze i1 %#3
  ret i1 %.fr
}
Transformation doesn't verify!

ERROR: Value mismatch

Example:
i16 %#0 = poison
i16 %#1 = #x0000 (0)

Source:
i1 %#3 = poison
i16 %#4 = poison
i16 %#5 = #x0000 (0)
i1 %#6 = #x1 (1)

Target:
i1 %#3 = poison
i1 %.fr = #x0 (0)
Source value: #x1 (1)
Target value: #x0 (0)

Summary:
  0 correct transformations
  1 incorrect transformations
  0 failed-to-prove transformations
  0 Alive2 errors
```

File: llvm/lib/Analysis/InstructionSimplify.cpp (hunk 1 above)
Comment (dtcxzyw) [reply llvm/lib/Analysis/InstructionSimplify.cpp]: Done.

**Review by nikic — APPROVED** [review]
LGTM

**Comment by tstellar** [comment]
@dtcxzyw Did you create a pull request for backporting this?

**Pull request merged**